File: 5842-programming-style/Assignment1-monolithic-freeflow/javascript/minesweeper.js
const readline = require('readline');
const fs = require('fs');
const rowEndings = "";
const fileEndings= ".cfg";

//constant of token of table drawings
const ttopleft = '\u250C'; //┌
const ttopmid = '\u252C'; //┬
const ttopright = '\u2510'; //┐
const tmidleft = '\u251C'; //├
const tmidmid = '\u253C'; //┼
const tmidright = '\u2524'; //┤
const tbotleft = '\u2514'; //└
const tbotmid = '\u2534'; //┴
const tbotright = '\u2518'; //┘
const thorizontal = '\u2500'; //─
const tvertical = '\u2502'; //│

//constant of token of message table
const mtopleft = '\u2554'; //╔
const mtopmid = '\u2550'; //═
const mtopright = '\u2557'; //╗
const mmidleft = '\u2551'; //║
const mmidright = '\u2551'; //║
const mbotleft = '\u255A'; //╚
const mbotmid = '\u2550'; //═
const mbotright = '\u255D'; //╝

const mine = '*';
const flag = '¶';
const empty = '▓';

const msgnotvalid = "The provided input is not valid!";
const msgwon = "You Won!";
const msglost = "You Lost!";
const msgnotenough = "Not enough inputs!";
const validinput = ['F', 'R'];
const maxColumnSize = 20;
var rowNumber =0;
var colNumber =0;
var message = "";
var input ="";
var gameOver = false;
var mineNumber = 0;
var openCellNumber = 0;


/** ALESSIO: Naming of functions that are not constructores shoud be in lowercase. */
function Cell( row, column, opened, flagged, mined, neighborMineCount ) 
{
	return {
		id: row + "_" + column,
		row: row,
		column: column,	
		opened: opened,
		flagged: flagged,
		mined: mined,
		neighborMineCount: neighborMineCount
	}
}

var checkBoard = function(fileName)
{
    // 1. check file existence and extension
    // file name have higher priority than file existence.
    var extension = fileName.toLowerCase().slice(fileName.length - fileEndings.length);
    var file = fileName.slice(0, fileName.length - fileEndings.length)
    if (extension != fileEndings || file.length ==0)
    {
        process.exitCode = 2;
        return 
    }    
    try{
	    fs.accessSync(fileName, fs.F_OK);    
    }
    catch (err){
        process.exitCode = 1;
        return
    }
    // 2. check board validity.
    try
    {
        const data = fs.readFileSync(fileName, 'utf-8'); 	  
		// Converting Raw Buffer to text 
        // data using tostring function. 
        var lines;
        if (data.endsWith('\n') == true){
            lines = data.toString().split("\n");     
            lines.pop(); //remove last element
        }
        else{
            process.exitCode =2;
            return
        }
        //console.log("lines.length:", lines.length);
        var columnsize = lines.length;
        var rowSize = 0;
        // A. empty file, return 2
        if (columnsize < 1 || columnsize > maxColumnSize)
        {
            //console.log("ohhh");
            process.exitCode = 2; 
            return
        }
        else{
            process.exitCode = 0;
            //console.log("ohhhhh");
            for(i in lines) 
            {
                regex = /^([\*\.])+$/g;
                var endings = lines[i].slice(lines[i].length - rowEndings.length);
                var row = lines[i].slice(0, lines[i].length - rowEndings.length);
                if (i==0) {rowSize = row.length;}
                if (/*endings != rowEndings ||*/ row.length != rowSize || (!(row.match(regex))))
                {
                    // not end with [\n] and row length != column size, return 2
                    process.exitCode = 2;
                    return
                }
            }
        }
        if ((columnsize == 1 && rowSize ==1) ||  rowSize > maxColumnSize)
        {
            process.exitCode = 2;
            return
        }
        //All checking asserted(except no. of mines), create board.
        board = Board(lines);
        return board;
    }
    catch (err){
        process.exitCode = 1;
        return
    }
}

function Board(lines)
{
	var board = {};
    //Read lines and count rowNumber and countNumber
	for(rowNumber = 1; rowNumber <= lines.length; rowNumber++ )
	{
		for(colNumber = 1; colNumber <= lines[rowNumber-1].length - rowEndings.length; colNumber++ )
		{
            if (lines[rowNumber-1][colNumber-1] == ".")
            {
                board[rowNumber + "_" + colNumber] = Cell( rowNumber, colNumber, false, false, false, 0 );
            }
            else
            {
                board[rowNumber + "_" + colNumber] = Cell( rowNumber, colNumber, false, false, true, 0 );
                mineNumber++;
            }
		}
    }
    rowNumber--;
    colNumber--;
    if (mineNumber >= rowNumber*colNumber)
    {
        process.exitCode = 2;
        return
    }
    
	board = calculateNeighborMineCounts( board, rowNumber, colNumber);
	return board;
}


var handleLeftClick = function(row, column)
{
    var cell = board[(row) + "_" + (column)];
    if( !cell.opened )
    {
        cell.flagged = false;
        if( cell.mined )
        {
            gameOver = true;
        }
        else
        {
            cell.opened = true;
            openCellNumber++;
            if( cell.neighborMineCount == 0 )
            {

                var neighbors = getNeighbors((row) + "_" + (column));
                for( var i = 0; i < neighbors.length; i++ )
                {
                    var neighbor = neighbors[i];
                    if(  typeof board[neighbor] !== 'undefined' &&
                            !board[neighbor].flagged && !board[neighbor].opened )
                    {
                        handleLeftClick((neighbor.split('_')[0]), (neighbor.split('_')[1]));
                    }
                }

            }
        }
    }
    else
    {
        message = msgnotvalid;
    }
}

var handleRightClick = function(row, column)
{
    var cell = board[(row) + "_" + (column)];
    if( !cell.opened )
    {
        if( !cell.flagged ) // if cell is not flagged
        {
            cell.flagged = true;      
        }
        else if( cell.flagged ) // if cell is flagged
        {
            cell.flagged = false;
        }
    }
    else // show er msg if flag an open cell
    {
        //message = msgnotvalid;
    }
}

var calculateNeighborMineCounts = function( board, rowNumber, colNumber )
{
	var cell;
	var neighborMineCount = 0;
	for( var row = 1; row <= rowNumber; row++ )
	{
		for( var column = 1; column <= colNumber; column++ )
		{
			var id = row + "_" + column;
			cell = board[id];
			if( !cell.mined )
			{
				var neighbors = getNeighbors(id);
				neighborMineCount = 0;
				for( var i = 0; i < neighbors.length; i++ )
				{
					neighborMineCount += isMined( board, neighbors[i] );
				}
				cell.neighborMineCount = neighborMineCount;
			}
		}
	}
	return board;
}


var getNeighbors = function(id)
{
	var row = parseInt(id.split('_')[0]);
	var column = parseInt(id.split('_')[1]);
	var neighbors = [];
	neighbors.push( (row - 1) + "_" + (column - 1) );
	neighbors.push( (row - 1) + "_" + column );
	neighbors.push( (row - 1) + "_" + (column + 1) );
	neighbors.push( row + "_" + (column - 1) );
	neighbors.push( row + "_" + (column + 1) );
	neighbors.push( (row + 1) + "_" + (column - 1) );
	neighbors.push( (row + 1) + "_" + column );
	neighbors.push( (row + 1) + "_" + (column + 1) );

	return neighbors
}


var isMined = function( board, id )
{	
	var cell = board[id];
	var mined = 0;
	if( typeof cell !== 'undefined' )
	{
		mined = cell.mined ? 1 : 0;
	}
	return mined;
}

// ALESSIO: This does not seem to make any sense? You implemented a function of one line, that creates a global variable
// validBoard, that is also returned by the function?
var ReadCfg = function(fileName)
{
  // TO-DO read .cfg files
  // create class board by reading files

	validBoard = checkBoard(fileName);
	return validBoard;
}

var DrawCell = function(cell)
{
    if (cell.flagged){ return flag;}
    // in displaying final mine map, this line have to come first to avoid bug.
    if (cell.opened && cell.mined) return mine;
    if (cell.opened && cell.neighborMineCount ==0) return empty;
    if (cell.opened && cell.neighborMineCount >0) return cell.neighborMineCount.toString();

    return " ";
}

var DrawBoard = function(board){
    // draw first row
    process.stdout.write(ttopleft);
    for (var col =1; col < colNumber; col++)
    {          
        process.stdout.write(thorizontal.repeat(3) + ttopmid);
    }
    process.stdout.write(thorizontal.repeat(3) + ttopright + rowEndings + '\n');
    // draw datarow + middle row
    for (var row =1; row <= rowNumber; row++)
    {
        // data row
        process.stdout.write(tvertical);
        for( var col = 1; col <= colNumber; col++ )
		{
            var id = row + "_" + col;
            cell = board[id];
            process.stdout.write(" " + DrawCell(cell) + " " + tvertical );
        }
        process.stdout.write(rowEndings+ '\n');
        if (row< rowNumber)
        {
            // middle row
            process.stdout.write(tmidleft);
            for (var col =1; col < colNumber; col++)
            {
                process.stdout.write(thorizontal.repeat(3) + tmidmid);
            }
            process.stdout.write(thorizontal.repeat(3) + tmidright+ rowEndings+ '\n');
        }
        else
        {
            // draw bottom row
            process.stdout.write(tbotleft);
            for (var col =1; col < colNumber; col++)
            {
                process.stdout.write(thorizontal.repeat(3) + tbotmid);
            }
            process.stdout.write(thorizontal.repeat(3) + tbotright + rowEndings + '\n');
        }
    }

}

var DrawMessage = function(colNumber)
{
    //determine size of column, compare colNumber and Message length
    messageLength = (message.length > colNumber*4-1) ? message.length : colNumber *4-1;
    additionalSpace = (colNumber*4-1-message.length <0 )? 0: colNumber*4-1-message.length;
    // draw first row
    process.stdout.write(mtopleft + mtopmid.repeat(messageLength) + mtopright+ rowEndings+ '\n');
    // draw message row
    process.stdout.write(mmidleft + message + " ".repeat(additionalSpace)+ mmidright+ rowEndings+ '\n');
    // draw bottom row
    process.stdout.write(mbotleft + mbotmid.repeat(messageLength) + mbotright+ rowEndings+ '\n');
}


var ParseInput = function(answer)
{
    // parse input
    var inputs = answer.split(/(\s+)/).filter( e => e.trim().length > 0)
    //console.log(inputs.length)
    if (inputs.length !=3) 
    {
        message = msgnotvalid;
    }
    else
    {
        //console.log(inputs[0] , inputs[1], inputs[2])
        //console.log((validinput.indexOf(inputs[2]) < 0))
        // check all invalid input of test
        if (inputs[0] < 1 || inputs[1] < 1 || inputs[0]> rowNumber || inputs[1] > colNumber|| 
            (validinput.indexOf(inputs[2]) < 0) || (inputs[0].indexOf('.') ==0) || (inputs[1].indexOf('.')==0))
        {
            message = msgnotvalid;
        }
        else{          
            message = ""; //reset the message.
            if (inputs[2] == "R")
            {
                handleLeftClick(inputs[0], inputs[1]);
            }
            else if (inputs[2] == "F")
            {
                handleRightClick(inputs[0], inputs[1]);
            }
        }
        //check if winning state is reach.
        if (mineNumber + openCellNumber == rowNumber * colNumber)
        {
            gameOver = true;
        }
    }
} 



var DrawFinalResult = function(msg)
{
    //console.clear()
    if (mineNumber + openCellNumber == rowNumber * colNumber) //winning state
    {
        message = msgwon;
    }
    else if (msg == msgnotenough)
    {
        message = msg;
    }
    else //losing state
    {
        message = msg;
        for( var row = 1; row <= rowNumber; row++ )
        {
            for( var column = 1; column <= colNumber; column++ )
            {
                var id = row + "_" + column;
                if (isMined(board, id)) {
                    cell = board[id];
                    cell.opened = true;
                    }
            }
        }
    }
    DrawBoard(board);
    DrawMessage(colNumber);
    process.exitCode = 0;
}
// var rl = readline.createInterface({
//     input: process.stdin,
//     output: process.stdout
// });

var InputPanel = function (rl) {
    //get answers.
    // 2. draw sweeper board
    process.exitCode = 0;
    DrawBoard(board);
    DrawMessage(colNumber);

    
    // 3a. wait for input
    rl.question('>', function (answer) {      
        // 3b. parse input
        ParseInput(answer);
        if (gameOver == true) //we need some base case, for recursion
        {
            // gameOver can be win or lost.
            DrawFinalResult(msglost);
            rl.close();
            return process.exitCode; //closing RL and returning from function.
        }
        InputPanel(rl); //Calling this function again to ask new question
    })
    
};

var main = function(args) 
{
	// Implement me !
    // 1. read file
    //console.log(args.length)
    if (args.length  < 3){
        process.exitCode = 1;
        /** ALESSIO: Fair, but you could also have used process.exit(exitCode) */
        return process.exitCode
    }

    /** ALESSIO: Does this also the validation? */
    board = ReadCfg(args[2]);
    
    if (process.exitCode === 0)
    {
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        InputPanel(rl);
        rl.on('close', function(answer) {
            // EOF
            if (gameOver != true){
                message = msgnotenough;
                DrawFinalResult(msgnotenough);
            }
            return process.exitCode;
           });
        //draw board for the last time.
        gameOver = false;
    }
    //return
    //console.log("exitcode: ", process.exitCode)
    return process.exitCode
}

if (require.main === module) {
	// Pass the arguments to the function
    // argv[0] -> node
    // argv[1] -> this script
    // argv[2] -> the first real input
    main(process.argv);
}

/*
  Export the main function so it can be called from the tests. This let us testing the
  app without creating a subprocess but YOU need to take care of cleaning up and
  initializing the app.
*/
module.exports.main = main

File: 5842-programming-style/Assignment1-monolithic-freeflow/java/Minesweeper.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class Minesweeper {

    public static void main(String[] args) {
        // declare chess-board property having dimension of n*m
        final String fileEndings = ".cfg";
        
        /** ALESSIO: Use the proper capitalization for variable names! */
        final String msgnotvalid = "The provided input is not valid!";
        final String msgwon = "You Won!";
        final String msglost = "You Lost!";
        final String msgnotenough = "Not enough inputs!";
        
        /** ALESSIO: What about maxRowSize? */
        final Integer maxColumnSize = 20;

        String message = "";
        String chessboard = "";
        Integer mineNumber = 0;
        // Integer openCellNumber = 0;
        Boolean gameOver = false;
        Integer rowNumber = 0;
        Integer colNumber = 0;
        // declare arrays below only after reading the file.
        // boolean[] flagged
        // bollean[] mined
        // boolean[] opened
        // int[] neightbourMineCount

        // And System.exit() is also the only way to specify the return value:
        // 1. read file
        // 1a. check if file name is valid

        // A possible violation of style, as substring are abstract methods.
        /** ALESSIO: Ok. */
        String fileFullName = args[0];
        String extension = fileFullName.toLowerCase().substring(fileFullName.length() - fileEndings.length());
        String fileName = fileFullName.substring(0, fileFullName.length() - fileEndings.length());
        // System.out.println("length: " + fileName.length());
        if (fileName.length() == 0 || !extension.equals(fileEndings)) {
            System.exit(2);
        }
        /** 
            ALESSIO: Since you call exit before, why not simply use a sequence of if-then-statement instead of nesting
            them?
        */
        // 1b. check if file exists
        else {
            try {
                File f = new File(fileFullName);
                if (!f.exists()) {
                    System.exit(1);
                } else if (f.isDirectory()) {
                    System.exit(2);
                } else {
                    // Possible violation of style, use Buffered Reader to read files and have IO
                    // exception.
                    /** ALESSIO: Ok */

                    /** 
                        ALESSIO: Why do you need a RandomAccessFile? that's not really necessary so 
                        I need to consider this a violation of the style! 
                        
                        What is this code about? To check if the file is empty was not enough to call 
                            File.size() == 0 ?
                    */
                    RandomAccessFile fileHandler = new RandomAccessFile(f, "r");

                    long fileLength = fileHandler.length() - 1;
                    if (fileLength < 0) {
                        fileHandler.close();
                        System.exit(2);
                    }
                    fileHandler.seek(fileLength);
                    byte readByte = fileHandler.readByte();
                    fileHandler.close();
                    if (!(readByte == 0xA)) {
                        System.exit(2);
                    }

                    chessboard = "";

                    // re read the file with while counting lines.
                    /** ALESSIO: This is not really smart, why reading the file twice when you could do it in one pass? */
                    Scanner myReader = new Scanner(f);
                    int tempColSize = 0;
                    while (myReader.hasNextLine()) {
                        String data = myReader.nextLine();
                        chessboard += data;
                        colNumber = data.length();
                        // 1c. check if every row have the same character.
                        if (rowNumber == 0) {
                            tempColSize = colNumber;
                            if (colNumber > maxColumnSize || colNumber < 1)
                                System.exit(2);
                        } else {
                            // 1d. check the length of each row.
                            if (tempColSize != colNumber || colNumber > maxColumnSize || colNumber < 1) {
                                System.exit(2);
                            }
                        }
                        rowNumber++;
                    }
                    myReader.close();


                    // 1e. check the length of total column & empty board.
                    if (rowNumber > maxColumnSize || rowNumber < 1 || chessboard.length() == 0) {
                        System.exit(2);
                    }
                    if (rowNumber == 1 && colNumber == 1) {
                        System.exit(2);
                    }
                    // 1f. check if all characters are * and .
                    /** ALESSIO: This is ok, but I cannot understand why you could not check this while reading each line */
                    for (int i = 0; i < chessboard.length(); i++) {
                        if (!(chessboard.charAt(i) == '*' || chessboard.charAt(i) == '.')) {
                            System.exit(2);
                        }
                    }
                }

            } catch (FileNotFoundException e) {
                System.exit(2);
            } catch (IOException e) {
                System.exit(2);
            }


            // 1g. put chessboard into arrays.
            /** ALESSIO: I cannot understand why not using a 2D array instead of a linear one... */
            // **************************************//
            boolean[] flagged = new boolean[chessboard.length()];
            boolean[] mined = new boolean[chessboard.length()];
            boolean[] opened = new boolean[chessboard.length()];
            int[] neighbourMineCount = new int[chessboard.length()];
            // **************************************//
            for (int i = 0; i < chessboard.length(); i++) {
                flagged[i] = false;
                opened[i] = false;
                mined[i] = (chessboard.charAt(i) == '*') ? true : false;
                if (chessboard.charAt(i) == '*')
                    mineNumber++;
            }
            // check number of mine.
            if (mineNumber == rowNumber * colNumber)
                System.exit(2);
            
            // 1h. calculate neightbourMineCount
            for (int i = 0; i < chessboard.length(); i++) {
                int minecount = 0;
                if (i % colNumber != 0) { // not left most column
                    if (i - colNumber - 1 > 0 && mined[i - colNumber - 1] == true)
                        minecount++;
                    if (mined[i - 1] == true)
                        minecount++;
                    if (i + colNumber - 1 < (rowNumber * colNumber) && mined[i + colNumber - 1] == true)
                        minecount++;
                }
                if (i - colNumber > 0 && mined[i - colNumber] == true)
                    minecount++;
                if (i + colNumber < (rowNumber * colNumber) && mined[i + colNumber] == true)
                    minecount++;
                if ((i + 1) % colNumber != 0) { // not right most column
                    if (i - colNumber + 1 > 0 && mined[i - colNumber + 1] == true)
                        minecount++;
                    if (mined[i + 1] == true)
                        minecount++;
                    if (i + colNumber + 1 < (rowNumber * colNumber) && mined[i + colNumber + 1] == true)
                        minecount++;
                }
                neighbourMineCount[i] = minecount;
            }


            Scanner sc = new Scanner(System.in);
            // 2. Draw Board + Process Result
            /** ALESSIO: This is inaccurate. While using whle true when you can have sc.hasNextLine() ?! */
            while (true) {
                // 2a. Draw Board
                // print top row
                System.out.print('┌');
                for (var col = 1; col < colNumber; col++) {
                    System.out.print("───┬");
                }
                System.out.println("───┐");

                // draw datarow + middle row
                for (var row = 1; row <= rowNumber; row++) {
                    // data row
                    System.out.print('│');
                    for (var col = 1; col <= colNumber; col++) {
                        var index = (row - 1) * colNumber + (col - 1);
                        System.out.print(" ");
                        // 2aI. print cell logic
                        if (flagged[index] == true)
                            System.out.print("¶");
                        else {
                            if (opened[index] == true && mined[index] == true)
                                System.out.print("*");
                            else if (opened[index] == true && neighbourMineCount[index] == 0)
                                System.out.print("▓");
                            else if (opened[index] == true && neighbourMineCount[index] > 0)
                                /** ALESSIO: Probably String.valueOf is unecessary */
                                System.out.print(String.valueOf(neighbourMineCount[index]));
                            else if (opened[index] == false)
                                System.out.print(" ");
                        }
                        System.out.print(" ");
                        System.out.print('│');
                    }

                    System.out.println();
                    if (row < rowNumber) {
                        // middle row
                        System.out.print('├');
                        for (var col = 1; col < colNumber; col++) {
                            System.out.print("───" + '┼');
                        }
                        System.out.println("───" + '┤');
                    } else {
                        // draw bottom row
                        System.out.print('└');
                        for (var col = 1; col < colNumber; col++) {
                            System.out.print("───" + '┴');
                        }
                        System.out.println("───" + '┘');
                    }
                }



                // 2b. Draw Message
                Integer messageLength = (message.length() > colNumber * 4 - 1) ? message.length() : colNumber * 4 - 1;
                Integer additionalSpace = (colNumber * 4 - 1 - message.length() < 0) ? 0
                        : colNumber * 4 - 1 - message.length();
                // draw first row
                System.out.print('╔');
                for (int cnt = 0; cnt < messageLength; cnt++)
                    System.out.print('═');
                System.out.println('╗');
                // draw message row
                System.out.print('║' + message);
                for (int cnt = 0; cnt < additionalSpace; cnt++)
                    System.out.print(' ');
                System.out.println('║');
                // draw bottom row
                System.out.print('╚');
                for (int cnt = 0; cnt < messageLength; cnt++)
                    System.out.print('═');
                System.out.println('╝');

                // 2c. Await Answer
                System.out.print(">");
                String input = "";
                message = "";
                if (sc.hasNextLine() == true)
                    input = sc.nextLine() + '\n';
                else {
                    message = msgnotenough;
                    gameOver = true;
                }

                // 2d. Parse Input
                // 2dI. parse to 3 inputs.
                if (gameOver != true) {
                    String[] inputs = { "", "", "", "" };
                    int inputsidx = 0;
                    int startchar = -1;
                    /** ALESSIO: I cannot understand why you did this. One one side, you want to parse the input by
                    character (to not violate the style), and then you use substring that might violated it? */
                    for (int idx = 0; idx < input.length(); idx++) {
                        if (inputsidx >= 4) {
                            break;
                        }
                        char ch = input.charAt(idx);
                        if (startchar == -1) { // We found the start of a word
                            if (ch == 'R' || ch == 'F' || (ch >= '0' && ch <= '9')) {
                                startchar = idx;
                            }
                            continue;
                        } else if (ch == ' ' || ch == '\n') { // We found the end of a word. Process it
                            // A possible violation of style, as substring are abstract methods.
                            if (startchar != -1) {
                                inputs[inputsidx] = input.substring(startchar, idx);
                                inputsidx++;
                            }
                            startchar = -1;
                            continue;
                        } else {
                            message = msgnotvalid;
                            break;
                        }
                    }


                    // 2dII. check if inputs[] are valid
                    // check if there are exactly 3 parameter, and if 3rd parameter is F and R.
                    if (inputs[3] != "" || !((inputs[2].equals("F")) || inputs[2].equals("R"))) {
                        message = msgnotvalid;
                    } else {
                        // check if 1st para and 2nd para are numbers only.
                        for (int idx = 0; idx < inputs[0].length(); idx++) {
                            char ch = inputs[0].charAt(idx);
                            if (!(ch >= '0' && ch <= '9'))
                                message = msgnotvalid;
                        }
                        for (int idx = 0; idx < inputs[1].length(); idx++) {
                            char ch = inputs[1].charAt(idx);
                            if (!(ch >= '0' && ch <= '9'))
                                message = msgnotvalid;
                        }
                    }

                    // 2e. Handle left-click
                    if (message != msgnotvalid && (inputs[2].equals("R"))) {
                        // message = "";
                        Integer idx = (Integer.parseInt(inputs[0]) - 1) * colNumber + Integer.parseInt(inputs[1]) - 1;
                        // System.out.println("R" + idx);
                        if (opened[idx] == false) {
                            if (mined[idx] == true) {
                                gameOver = true;
                            } else {
                                opened[idx] = true;
                                if (neighbourMineCount[idx] == 0) {
                                    boolean openAllCells = false;
                                    while (!openAllCells) {
                                        // break statement: all cell with open empty neighbour are opened already
                                        openAllCells = true;
                                        for (int i = 0; i < chessboard.length(); i++) {
                                            // check only cells which is not opened
                                            if (opened[i] == false) {
                                                // not left most column
                                                if (i % colNumber != 0) {
                                                    if (i - colNumber - 1 > 0 && opened[i - colNumber - 1] == true
                                                            && neighbourMineCount[i - colNumber - 1] == 0)
                                                        openAllCells = false;
                                                    if (opened[i - 1] == true && neighbourMineCount[i - 1] == 0)
                                                        openAllCells = false;
                                                    if (i + colNumber - 1 < (rowNumber * colNumber)
                                                            && opened[i + colNumber - 1] == true
                                                            && neighbourMineCount[i + colNumber - 1] == 0)
                                                        openAllCells = false;
                                                }

                                                if (i - colNumber > 0 && opened[i - colNumber] == true
                                                        && neighbourMineCount[i - colNumber] == 0)
                                                    openAllCells = false;
                                                if (i + colNumber < (rowNumber * colNumber)
                                                        && opened[i + colNumber] == true
                                                        && neighbourMineCount[i + colNumber] == 0)
                                                    openAllCells = false;

                                                // not right most column
                                                if ((i + 1) % colNumber != 0) {
                                                    if (i - colNumber + 1 > 0 && opened[i - colNumber + 1] == true
                                                            && neighbourMineCount[i - colNumber + 1] == 0)
                                                        openAllCells = false;
                                                    if (opened[i + 1] == true && neighbourMineCount[i + 1] == 0)
                                                        openAllCells = false;
                                                    if (i + colNumber + 1 < (rowNumber * colNumber)
                                                            && opened[i + colNumber + 1] == true
                                                            && neighbourMineCount[i + colNumber + 1] == 0)
                                                        openAllCells = false;
                                                }
                                            }
                                        }
                                        // generic statement: get the chessboard
                                        if (openAllCells == false) {
                                            for (int i = 0; i < chessboard.length(); i++) {
                                                // check for all opened + 0 neighbourMineCount cell
                                                if (opened[i] == true && neighbourMineCount[i] == 0) {
                                                    // open their all neighbours
                                                    if (i % colNumber != 0) {
                                                        // top left
                                                        if (i - colNumber - 1 > 0)
                                                            opened[i - colNumber - 1] = true;
                                                        // mid left
                                                        opened[i - 1] = true;
                                                        // bot left
                                                        if (i + colNumber - 1 < (rowNumber * colNumber))
                                                            opened[i + colNumber - 1] = true;
                                                    }
                                                    // top mid
                                                    if (i - colNumber > 0)
                                                        opened[i - colNumber] = true;
                                                    // bot mid
                                                    if (i + colNumber < (rowNumber * colNumber))
                                                        opened[i + colNumber] = true;

                                                    if ((i + 1) % colNumber != 0) {
                                                        // top right
                                                        if (i - colNumber + 1 > 0)
                                                            opened[i - colNumber + 1] = true;
                                                        // mid right
                                                        opened[i + 1] = true;
                                                        // bottom right
                                                        if (i + colNumber + 1 < (rowNumber * colNumber))
                                                            opened[i + colNumber + 1] = true;
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                                // check open cell number
                                Integer openCellNumber = 0;
                                for (int i = 0; i < chessboard.length(); i++) {
                                    if (opened[i] == true)
                                        openCellNumber++;
                                }
                                if (mineNumber + openCellNumber == rowNumber * colNumber) {
                                    gameOver = true;
                                }
                            }
                        }

                    }
                    // 2f. Handle right-click
                    if (message != msgnotvalid && (inputs[2].equals("F"))) {
                        // message = "";
                        /** ALESSIO: This is another (minor) violation of the style: parseInt */
                        Integer idx = (Integer.parseInt(inputs[0]) - 1) * colNumber + Integer.parseInt(inputs[1]) - 1;
                        // System.out.println("F" + idx);
                        if (opened[idx] == false) {
                            if (flagged[idx] == true) {
                                flagged[idx] = false;
                            } else {
                                flagged[idx] = true;
                            }
                        }
                    }
                }
                // 2g. check win/lose/not enough input state.
                if (gameOver == true) {
                    Integer openCellNumber = 0;
                    for (int i = 0; i < chessboard.length(); i++) {
                        if (opened[i] == true)
                            openCellNumber++;
                    }
                    // System.out.println("message " + message);
                    if (mineNumber + openCellNumber == rowNumber * colNumber) {
                        message = msgwon;
                    } else if (message.equals(msgnotenough)) {
                        message = msgnotenough;
                    } else {
                        message = msglost;
                    }
                    // open all cells with mines
                    if (message.equals(msglost)) {
                        for (int i = 0; i < chessboard.length(); i++) {
                            if (mined[i] == true)
                                opened[i] = true;
                        }
                    }
                    // Draw message board and chess board for last time.
                    // 2gI. Draw Board
                    // print top row
                    System.out.print('┌');
                    for (var col = 1; col < colNumber; col++) {
                        System.out.print("───┬");
                    }
                    System.out.println("───┐");
                    // draw datarow + middle row
                    for (var row = 1; row <= rowNumber; row++) {
                        // data row
                        System.out.print('│');
                        for (var col = 1; col <= colNumber; col++) {
                            var index = (row - 1) * colNumber + (col - 1);
                            System.out.print(" ");
                            // 2aI. print cell logic
                            if (flagged[index] == true)
                                System.out.print("¶");
                            else {
                                if (opened[index] == true && mined[index] == true)
                                    System.out.print("*");
                                else if (opened[index] == true && neighbourMineCount[index] == 0)
                                    System.out.print("▓");
                                else if (opened[index] == true && neighbourMineCount[index] > 0)
                                    System.out.print(String.valueOf(neighbourMineCount[index]));
                                else if (opened[index] == false)
                                    System.out.print(" ");
                            }
                            System.out.print(" ");
                            System.out.print('│');
                        }
                        System.out.println();
                        if (row < rowNumber) {
                            // middle row
                            System.out.print('├');
                            for (var col = 1; col < colNumber; col++) {
                                System.out.print("───" + '┼');
                            }
                            System.out.println("───" + '┤');
                        } else {
                            // draw bottom row
                            System.out.print('└');
                            for (var col = 1; col < colNumber; col++) {
                                System.out.print("───" + '┴');
                            }
                            System.out.println("───" + '┘');
                        }
                    }
                    // 2gII. Draw Message
                    messageLength = (message.length() > colNumber * 4 - 1) ? message.length() : colNumber * 4 - 1;
                    additionalSpace = (colNumber * 4 - 1 - message.length() < 0) ? 0
                            : colNumber * 4 - 1 - message.length();
                    // draw first row
                    System.out.print('╔');
                    for (int cnt = 0; cnt < messageLength; cnt++)
                        System.out.print('═');
                    System.out.println('╗');
                    // draw message row
                    System.out.print('║' + message);
                    for (int cnt = 0; cnt < additionalSpace; cnt++)
                        System.out.print(' ');
                    System.out.println('║');
                    // draw bottom row
                    System.out.print('╚');
                    for (int cnt = 0; cnt < messageLength; cnt++)
                        System.out.print('═');
                    System.out.println('╝');

                    // close input channel
                    sc.close();
                    break;
                }
            }
        }
        // System.exit(0);
    }

}